ic-canister-client/src/state_machine_tests.rs

```diff
@@ -0,0 +1,121 @@
+use std::sync::Arc;
+
+use candid::utils::ArgumentEncoder;
+use candid::{CandidType, Decode, Principal};
+use ic_exports::ic_kit::RejectionCode;
+use ic_exports::ic_test_state_machine::{StateMachine, WasmResult};
+use serde::Deserialize;
+use tokio::sync::Mutex;
+
+use crate::{CanisterClient, CanisterClientError, CanisterClientResult};
+
+/// A client for interacting with a canister inside dfinity's
+/// state machine tests framework.
+#[derive(Clone)]
+pub struct StateMachineCanisterClient {
+    state_machine: Arc<Mutex<StateMachine>>,
+    canister: Principal,
+    caller: Principal,
+}
+
+impl StateMachineCanisterClient {
+    /// Creates a new instance of a StateMachineCanisterClient.
+    pub fn new(
+        state_machine: Arc<Mutex<StateMachine>>,
```

Comment:
we do you need an async mutex?

Reply:
Async mutex needed because the client provides async API.

Reply:
maybe I am wrong but I remember that StateMachine client methods are not mutable, so `Arc<StateMachine>` should be enough

Reply:
StateMachine is !Sync.

Reply:
I don't get it. Could you please add a simple integration tests that uses the state machine client?